torch/utils/_foreach_utils.py

```diff
@@ -39,7 +39,9 @@ def _group_tensors_by_device_and_dtype(
     }
 
 
+def _device_has_foreach_support(device: torch.device) -> bool:
+    return device.type in (_get_foreach_kernels_supported_devices() + ["cpu"]) and not torch.jit.is_scripting()
```

Comment:
I think converting to set takes a linear pass over the list, so we might as well just keep it as a list.